### DISCUSSION TITLE
Why the Permissions Matrix not stored in db

### DISCUSSION THREAD

**Discussion by mshannaq** [codeigniter4/shield#805] [Q&A]
As I can see the group permissions is stored in `app/Config/AuthGroups.php` in `array $matrix`

that mean the group permissions is static and only can be changed by updating the file `AuthGroups.php`. and this prevent the users from changing the group permissions unless they access the file system.

what if we need to give the ability to logged in super users to modify the group permissions? I think that require the group permissions to be stored in DB. is that correct?

**Comment by datamweb** [comment]
Shield uses the official library [CodeIgniter 4 Settings](https://github.com/codeigniter4/settings).
So you can set your preferences using it.
Please read the library [documentation ](https://github.com/codeigniter4/settings#codeigniter-4-settings).

**Reply by mshannaq** [reply]
Yes , I got it

I added the matrix array into the table settings using

example:

```
$matrix = [
        'superadmin' => [
            'admin.*',
            'users.*',
            'beta.*',
        ],
        'admin' => [
            'admin.access',
            'users.create',
            'users.edit',
            'users.delete',
            'beta.access',
        ],
        'developer' => [
            'admin.access',
            'admin.settings',
            'users.create',
            'users.edit',
            'beta.access',
        ],
        'user' => [],
        'beta' => [
            'beta.access',
        ],
    ];

service('settings')->set('AuthGroups.matrix', $matrix);
```

Then Shiled will take the permissions of the groups from the database instead of file system (while setting AuthGroups.matrix  is exists on the settings table) 

![image](https://github.com/codeigniter4/shield/assets/162535/8e3561aa-4715-4451-900c-16b51df8f564)

as it defined in https://github.com/codeigniter4/shield/blob/b6d327d1cfdc4a492697f79174b0e159da6a12e3/src/Authorization/Traits/Authorizable.php#L262-L264
